src/FakeItEasy/Core/ArgumentEqualityComparer.cs

```diff
@@ -0,0 +1,37 @@
+﻿namespace FakeItEasy.Core
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    internal class ArgumentEqualityComparer
+    {
+        private readonly IEnumerable<IArgumentEqualityComparer> argumentEqualityComparers;
+        private readonly ConcurrentDictionary<Type, IArgumentEqualityComparer?> cachedComparers = new();
+
+        public ArgumentEqualityComparer(IEnumerable<IArgumentEqualityComparer> argumentEqualityComparers)
+        {
+            this.argumentEqualityComparers = argumentEqualityComparers.OrderByDescending(c => c.Priority).ToArray();
+        }
+
+        public bool AreEqual(object? expectedValue, object? argumentValue, Type parameterType)
+        {
+            var comparer = this.cachedComparers.GetOrAdd(parameterType, t => this.argumentEqualityComparers.FirstOrDefault(c => c.CanCompare(t)));

```

Comment:
It's been so long since I LINQed, I was surprised to see a `FirstOrDefault` that took a predicate. So many convenience methods. Such decadence. My mind was prepared for a filter and then taking first (or default).